File: app/src/main/java/nz/co/martinpaulo/wordjambalaya/Dictionary.java
package nz.co.martinpaulo.wordjambalaya;

import android.content.Context;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Lots of room for optimisation!
 */
public final class Dictionary {

    private static final String TAG = "Dictionary";

    HashMap<String, Set<String>> sortedWords = new HashMap<String, Set<String>>();

    private static Dictionary instance = new Dictionary();
    private boolean isInitialized;
    private String selectChars = "";

    private Dictionary() {
        // made private to force singleton pattern
    }

    private void addWord(final String word) {
        String key = orderCharacters(word);
        final Set<String> words = sortedWords.get(key);
        if (words == null) {
            sortedWords.put(key, new HashSet<String>(Arrays.asList(word)));
            return;
        }
        words.add(word);
    }

    protected String orderCharacters(final String word) {
        char[] content = word.toCharArray();
        java.util.Arrays.sort(content);
        return new String(content);
    }

    /**
     * @param jumbledWord The jumbled word to decode.
     * @return the set of words that match the jumbled word.
     */
    public List<String> findJumbledWord(final String jumbledWord) {
        List<String> answer = new ArrayList<String>();
        final Set<String> candidate = sortedWords.get(orderCharacters(jumbledWord));
        if (candidate != null) {
            answer.addAll(candidate);
        }
        return answer;
    }

    /**
     * @param length                            the length of the words that we are looking for
     * @param muddledWordsWithOrderedCharacters the set of muddled words, with characters ordered.
     * @return the list of words of the given length that are a subset of the ordered characters
     */
    private List<String> getCandidateWords(int length, final String muddledWordsWithOrderedCharacters) {
        List<String> result = new ArrayList<String>();
        nextWord:
        for (String key : sortedWords.keySet()) {
            if (key.length() == length) {
                for (int i = 0; i < key.length(); i++) {
                    if (muddledWordsWithOrderedCharacters.indexOf(key.charAt(i)) <= -1) {
                        continue nextWord;
                    }
                }
                result.addAll(sortedWords.get(key));
            }
        }
        return result;
    }

    private List<List<String>> getPossibleAnswers(final List<Integer> wordLengths,
                                                  final String muddledWordsWithOrderedCharacters) {
        List<List<String>> columns = new ArrayList<List<String>>();
        for (int wordLength : wordLengths) {
            columns.add(getCandidateWords(wordLength, muddledWordsWithOrderedCharacters));
        }

        return getCandidateWordSets(muddledWordsWithOrderedCharacters, columns);
    }

    protected String orderCharactersOf(final List<String> words) {
        String result = "";
        for (String word : words) {
            result = result + word;
        }
        return orderCharacters(result);
    }

    protected List<List<String>> getCandidateWordSets(String muddledWordsWithOrderedCharacters, List<List<String>> columns) {
        final ArrayList<List<String>> result = new ArrayList<List<String>>();
        recursivelyFindCandidateWordSets(muddledWordsWithOrderedCharacters, columns, result, 0, new ArrayList<String>());
        return result;
    }

    protected void recursivelyFindCandidateWordSets(String muddledWordsWithOrderedCharacters,
                                                    List<List<String>> columns,
                                                    List<List<String>> result,
                                                    final int currentColumn,
                                                    List<String> candidateWordSet) {
        for (String word : columns.get(currentColumn)) {
            List<String> newCandidateWordSet = new ArrayList<String>(candidateWordSet);
            newCandidateWordSet.add(word);
            if (currentColumn >= columns.size() - 1) {
                testAndAddToResultIfMatch(result, muddledWordsWithOrderedCharacters, newCandidateWordSet);
            } else {
                recursivelyFindCandidateWordSets(muddledWordsWithOrderedCharacters, columns, result, currentColumn + 1, newCandidateWordSet);
            }
        }
    }

    private void testAndAddToResultIfMatch(List<List<String>> result, String muddledWordsWithOrderedCharacters, List<String> candidateWords) {
        String orderedWord = orderCharactersOf(candidateWords);
        if (orderedWord.equals(muddledWordsWithOrderedCharacters)) {
            result.add(candidateWords);
        }
    }


    public List<List<String>> findPossibleAnswers(final List<Integer> wordLengths) {
        Log.i(TAG, "Looking for " + wordLengths.size() + " words formed from " + selectChars);
        final String orderedCharactersOfMuddledWords = orderCharacters(selectChars);
        return getPossibleAnswers(wordLengths, orderedCharactersOfMuddledWords);
    }

    /**
     * Need to make sure that this not reloaded if the configuration changes...
     */
    static Dictionary buildDictionary(Context ctx) {
        Class raw = R.raw.class;
        Field[] fields = raw.getFields();
        for (Field field : fields) {
            try {
                instance.readRawTextFile(ctx, field.getInt(null));
            } catch (IllegalAccessException e) {
                Log.e(TAG, String.format("%s threw IllegalAccessException.",
                        field.getName()));
            }
        }
        instance.isInitialized = true;
        return instance;
    }

    public static Dictionary getInstance() {
        if (!instance.isInitialized) {
            // throw new exception here?
        }
        return instance;
    }

    private void readRawTextFile(Context ctx, int resId) {
        InputStream inputStream = ctx.getResources().openRawResource(resId);
        InputStreamReader inputReader = new InputStreamReader(inputStream);
        BufferedReader bufferReader = new BufferedReader(inputReader, 8192);
        String line;
        try {
            while ((line = bufferReader.readLine()) != null) {
                addWord(line);
            }
        } catch (IOException e) {
            Log.e(TAG, "Could not read word resource.", e);
        }
    }

    public void addSelectedChar(char character) {
        selectChars = selectChars + character;
    }

    public void removeSelectedChar(char character) {
        selectChars = selectChars.replaceFirst(String.valueOf(character), "");
    }

    public Integer getAnswerLength() {
        return selectChars.length();
    }
}

File: app/src/main/java/nz/co/martinpaulo/wordjambalaya/Dictionary.kt
package nz.co.martinpaulo.wordjambalaya

import android.content.Context
import android.os.Process
import android.util.Log
import nz.co.martinpaulo.wordjambalaya.R.raw
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.util.Objects

/**
 * A singleton wrapper around the loaded word list.
 * The words are kept in a hashmap, keyed by the sorted list of characters in the words.
 * So, for example:
 * "abc" -> "cab"
 * "abd" -> "bad", "dab"
 * "abn" -> "ban", "nab"
 *
 */
class Dictionary private constructor() {
    private var sortedWords: HashMap<String, MutableSet<String>> = HashMap()

    private var isInitialized = false // has the dictionary been loaded?
    /** the list of characters selected to make up the riddle answer */
    private var riddleAnswerChars = ""

    private fun addWord(word: String) {
        val key = orderCharacters(word)
        val words = sortedWords[key]
        if (words == null) { // no entry for this set of characters
            // we use a set as the words being added on later runs might already be in the set
            sortedWords[key] = HashSet(listOf(word))
        } else {
            // add to the already existing set of words with matching ordered characters
            words.add(word)
        }
    }

    /**
     * Returns the word parameter with its characters sorted in alphabetic order.
     * For example: "hearty" -> "aehrty"
     * @param word the word whose characters are to be ordered
     * @return the input word, now with its characters sorted in alphabetic order
     */
    fun orderCharacters(word: String): String {
        return word.toCharArray().sorted().joinToString("")
    }

    /**
     * @param jumbledWord the jumbled word to decode.
     * @return the list of words that match the jumbled word (if any).
     */
    fun findJumbledWords(jumbledWord: String): List<String> {
        val answer: MutableList<String> = ArrayList()
        val candidates: Set<String>? = sortedWords[orderCharacters(jumbledWord)]
        if (candidates != null) {
            answer.addAll(candidates)
        }
        return answer
    }

    /**
     * @param length                            the length of the words that we are looking for
     * @param muddledWordsWithOrderedCharacters the set of muddled words, with the characters ordered.
     * @return the list of words of the given length that are a subset of the ordered characters
     */
    private fun getCandidateWords(
        length: Int,
        muddledWordsWithOrderedCharacters: String
    ): List<String> {
        val result: MutableList<String> = ArrayList()
        nextWord@ for (key in sortedWords.keys) {
            if (length == key.length) {
                for (element in key) {
                    if (muddledWordsWithOrderedCharacters.indexOf(element) == -1) {
                        continue@nextWord
                    }
                }
                result.addAll(Objects.requireNonNull<Set<String>>(sortedWords[key])
                )
            }
        }
        return result
    }

    /**
     *
     * @param wordLengths
     * @param muddledWordsWithOrderedCharacters
     * @return
     */
    private fun getPossibleAnswers(
        wordLengths: List<Int?>?,
        muddledWordsWithOrderedCharacters: String
    ): List<List<String>> {
        val columns: MutableList<List<String>> = ArrayList()
        if (wordLengths != null) {
            for (wordLength in wordLengths) {
                if (wordLength != null)
                    columns.add(getCandidateWords(wordLength, muddledWordsWithOrderedCharacters))
            }
        }
        return getCandidateWordSets(muddledWordsWithOrderedCharacters, columns)
    }

    /**
     *
     * @param words a list of words
     * @return the list of input words concatenated and then with their characters sorted
     */
    fun orderCharactersOf(words: List<String>): String {
        return orderCharacters(words.joinToString("") { it }.takeWhile { it.isLetter() })
    }

    /**
     *
     * @param muddledWordsWithOrderedCharacters
     * @param columns
     * @return
     */
    fun getCandidateWordSets(
        muddledWordsWithOrderedCharacters: String,
        columns: List<List<String>>
    ): List<List<String>> {
        val result = ArrayList<List<String>>()
        recursivelyFindCandidateWordSets(
            muddledWordsWithOrderedCharacters,
            columns,
            result,
            0,
            ArrayList()
        )
        return result
    }

    private fun recursivelyFindCandidateWordSets(
        muddledWordsWithOrderedCharacters: String,
        columns: List<List<String>>,
        result: MutableList<List<String>>,
        currentColumn: Int,
        candidateWordSet: List<String>
    ) {
        for (word in columns[currentColumn]) {
            val newCandidateWordSet: MutableList<String> = ArrayList(candidateWordSet)
            newCandidateWordSet.add(word)
            if (currentColumn >= columns.size - 1) {
                testAndAddToResultIfMatch(
                    result,
                    muddledWordsWithOrderedCharacters,
                    newCandidateWordSet
                )
            } else {
                recursivelyFindCandidateWordSets(
                    muddledWordsWithOrderedCharacters,
                    columns,
                    result,
                    currentColumn + 1,
                    newCandidateWordSet
                )
            }
        }
    }

    /**
     *
     * @param result
     * @param muddledWordsWithOrderedCharacters
     * @param candidateWords
     */
    private fun testAndAddToResultIfMatch(
        result: MutableList<List<String>>,
        muddledWordsWithOrderedCharacters: String,
        candidateWords: List<String>
    ) {
        val orderedWord = orderCharactersOf(candidateWords)
        if (orderedWord == muddledWordsWithOrderedCharacters) {
            result.add(candidateWords)
        }
    }


    fun findPossibleRiddleAnswers(wordLengths: List<Int?>?): List<List<String>> {
        Log.i(TAG, "Looking for " + (wordLengths?.size ?: 1) + " words formed from " + riddleAnswerChars)
        val orderedCharactersOfMuddledWords = orderCharacters(riddleAnswerChars)
        return getPossibleAnswers(wordLengths, orderedCharactersOfMuddledWords)
    }

    private fun readWordFile(ctx: Context, resId: Int) {
        val inputStream = ctx.resources.openRawResource(resId)
        val inputReader = InputStreamReader(inputStream)
        val bufferReader = BufferedReader(inputReader, 8192)
        var line: String
        try {
            while ((bufferReader.readLine().also { line = it }) != null) {
                addWord(line)
            }
        } catch (e: IOException) {
            Log.e(TAG, "Could not read word resource.", e)
        }
    }

    /**
     * Add a character to the list of characters to be used to make the riddle answer.
     */
    fun addCharToRiddleAnswer(character: Char) {
        riddleAnswerChars += character
    }

    /**
     * Remove the first instance of the character from the list of characters to be used to make the
     * riddle  answer.
     */
    fun removeCharFromRiddleAnswer(character: Char) {
        riddleAnswerChars = riddleAnswerChars.replaceFirst(character.toString(), "")
    }

    /**
     * The length of the answer string so far
     */
    val riddleAnswerLength: Int
        get() = riddleAnswerChars.length

    companion object {
        private const val TAG = "Dictionary"
        private val instance = Dictionary()

        /**
         * Need to make sure that this not reloaded if the configuration changes...
         */
        @JvmStatic
        fun buildDictionary(ctx: Context): Dictionary {
            val raw = raw::class.java   // represents the raw folder
            var exceptionThrown = false // record if there were any errors reading the files
            for (file in raw.fields) {  // for each of the files in the raw folder...
                try {
                    instance.readWordFile(ctx, file.getInt(null))
                } catch (e: IllegalAccessException) {
                    exceptionThrown = true
                    Log.e(TAG, String.format("%s threw IllegalAccessException.", file.name)
                    )
                }
            }
            instance.isInitialized = !exceptionThrown
            return instance
        }

        @JvmStatic
        fun getInstance(): Dictionary {
            if (!instance.isInitialized) {
                // throw new exception here rather than exit?
                val pid = Process.myPid()
                Process.killProcess(pid)
            }
            return instance
        }
    }
}
